### PR TITLE
Added support for selecting best candidate for some special string function

### PR DESCRIPTION
**Pull request by rohit01010**
### Description
This PR contains following changes:
- For implementing wrapper functions for TRIM function, an additional handling was required in selecting best candidate, used a hook `func_select_candidate_for_exception_hook` for the same.
- extended definition of `report_proc_not_found_error_hook ` to include `actual_arg_types`, which will be used to throw more specific error in case of exception function.
- In case of SQL dialect, used sys.TRIM function instead of Postgres function.

Extension PR: https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/2495

Authored-by: Rohit Bhagat [rohitbgt@amazon.com](mailto:rohitbgt@amazon.com)
Signed-off-by: Rohit Bhagat [rohitbgt@amazon.com](mailto:rohitbgt@amazon.com)

### Issues Resolved

BABEL-4489

### Check List

- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/backend/parser/gram.y [hunk 1]
```diff
@@ -15551,10 +15551,16 @@ func_expr_common_subexpr:
...
 			| TRIM '(' trim_list ')'
 				{
-					$$ = (Node *) makeFuncCall(SystemFuncName("btrim"),
-											   $3,
-											   COERCE_SQL_SYNTAX,
-											   @1);
+					if (sql_dialect == SQL_DIALECT_TSQL)
+						$$ = (Node *) makeFuncCall(list_make2(makeString("sys"), makeString("trim")),
+													$3,
+													COERCE_EXPLICIT_CALL,
```
Comment (Deepesh125) [inline src/backend/parser/gram.y]: what is meaning of this COERCE_EXPLICIT_CALL and COERCE_SQL_SYNTAX?

Comment (rohit01010) [reply src/backend/parser/gram.y]: `COERCE_EXPLICIT_CALL` means display `FuncExpr` as function call, whereas `COERCE_SQL_SYNTAX` means display `FuncExpr` with SQL-mandated special syntax.

File: src/backend/access/common/tupconvert.c [hunk 2]
```diff
@@ -164,7 +164,8 @@ execute_attr_map_tuple(HeapTuple tuple, TupleConversionMap *map)
 		/**
 		 * if it's tsql insert exec, we'll consider value cast
```
Comment (Anikait143) [inline src/backend/access/common/tupconvert.c]: NIT: Please Update the comment:
```if it's tsql insert exec``` -> ```if it's tsql insert exec or tsql inline table valued function```

File: src/backend/parser/parse_func.c [hunk 3]
```diff
@@ -1080,8 +1081,8 @@ func_select_candidate(int nargs,
 	 * If we resolve all the unknwon types but still too many candidates,
 	 * let's try to choose the best candidate by T-SQL precedence rule.
 	 */
-	if (nunknowns == 0 &&
-	    (sql_dialect == SQL_DIALECT_TSQL ||
+	if (nunknowns == 0 && 
+		(sql_dialect == SQL_DIALECT_TSQL ||
```
Comment (Anikait143) [inline src/backend/parser/parse_func.c]: NIT: unnecessary change

File: src/backend/access/common/tupconvert.c (hunk 2 above)
Comment (rohit01010) [reply src/backend/access/common/tupconvert.c]: updated

File: src/backend/parser/parse_func.c (hunk 3 above)
Comment (rohit01010) [reply src/backend/parser/parse_func.c]: removed

**Review by Anikait143 — APPROVED** [review]

File: src/backend/parser/gram.y [hunk 4]
```diff
@@ -15551,10 +15551,16 @@ func_expr_common_subexpr:
 				}
 			| TRIM '(' trim_list ')'
 				{
-					$$ = (Node *) makeFuncCall(SystemFuncName("btrim"),
-											   $3,
-											   COERCE_SQL_SYNTAX,
-											   @1);
+					if (sql_dialect == SQL_DIALECT_TSQL)
+						$$ = (Node *) makeFuncCall(list_make2(makeString("sys"), makeString("trim")),
+													$3,
```
Comment (shah-nirmit) [inline src/backend/parser/gram.y]: Do we need to really introduce engine parser change, can we explore some non-invasive options

Comment (rishabhtanwar29) [reply src/backend/parser/gram.y]: 1. +1 to above, can't we identify TRIM node in ANTLR and modify it to sys.TRIM even before it comes here.
2. Also, what about LEADING/TRAILING/BOTH variants of `TRIM`? don't we need to handle them as well?

File: src/backend/parser/gram.y (hunk 1 above)
Comment (rishabhtanwar29) [reply src/backend/parser/gram.y]: Why not keep the coercion form same as `COERCE_SQL_SYNTAX`? After-all, it is considered a special syntax in T-SQL as well.

File: src/include/parser/parse_func.h [hunk 5]
```diff
@@ -75,10 +75,11 @@ extern void check_srf_call_placement(ParseState *pstate, Node *last_srf,
  * Hook interface to select a function from candidates
  */
 typedef FuncCandidateList (*func_select_candidate_hook_type) (int nargs, Oid *input_typeids, FuncCandidateList candidates, bool unknowns_resolved);
+typedef FuncCandidateList (*func_select_candidate_for_special_func_hook_type) (List *names, int nargs, Oid *input_typeids, FuncCandidateList candidates);
```
Comment (rishabhtanwar29) [inline src/include/parser/parse_func.h]: We can possibly reuse the existing `func_select_candidate_hook` hook for the special functions as well isn't it? Just need to update the hook function to have an extra boolean argument to distinguish between normal and special case.

Comment (shah-nirmit) [reply src/include/parser/parse_func.h]: +1

File: src/backend/parser/gram.y (hunk 1 above)
Comment (rohit01010) [reply src/backend/parser/gram.y]: removed this check.

File: src/backend/parser/gram.y (hunk 4 above)
Comment (rohit01010) [reply src/backend/parser/gram.y]: 1. removed engine parser change
2. we currently do not support `LEADING/TRAILING/BOTH` in babelfish, this support will be added later separately.

File: src/include/parser/parse_func.h (hunk 5 above)
Comment (rohit01010) [reply src/include/parser/parse_func.h]: Done

**Review by rishabhtanwar29 — APPROVED** [review]

**Review by shah-nirmit — APPROVED** [review]

**Review by Deepesh125 — APPROVED** [review]
